// raindrops-cli/build/utils/pda.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEdition = exports.getMetadata = exports.getItemEscrow = exports.getCraftItemEscrow = exports.getCraftItemCounter = exports.getItemActivationMarker = exports.getPlayerPDA = exports.getItemPDA = exports.getNamespacePDA = exports.getOracle = exports.getMatchTokenAccountEscrow = exports.getMatch = exports.getAtaForMint = void 0;
const anchor_1 = require("@project-serum/anchor");
const programIds_1 = require("../constants/programIds");
const item_1 = require("../constants/item");
const matches_1 = require("../constants/matches");
const namespace_1 = require("../constants/namespace");
const player_1 = require("../constants/player");
const getAtaForMint = async (mint, wallet) => {
    return await anchor_1.web3.PublicKey.findProgramAddress([wallet.toBuffer(), programIds_1.TOKEN_PROGRAM_ID.toBuffer(), mint.toBuffer()], programIds_1.SPL_ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID);
};
exports.getAtaForMint = getAtaForMint;
const getMatch = async (oracle) => {
    return await anchor_1.web3.PublicKey.findProgramAddress([Buffer.from(matches_1.PREFIX), oracle.toBuffer()], programIds_1.MATCHES_ID);
};
exports.getMatch = getMatch;
const getMatchTokenAccountEscrow = async (oracle, tokenMint, tokenOwner) => {
    return await anchor_1.web3.PublicKey.findProgramAddress([
        Buffer.from(matches_1.PREFIX),
        oracle.toBuffer(),
        tokenMint.toBuffer(),
        tokenOwner.toBuffer(),
    ], programIds_1.MATCHES_ID);
};
exports.getMatchTokenAccountEscrow = getMatchTokenAccountEscrow;
const getOracle = async (seed, payer) => {
    return await anchor_1.web3.PublicKey.findProgramAddress([Buffer.from(matches_1.PREFIX), payer.toBuffer(), seed.toBuffer()], programIds_1.MATCHES_ID);
};
exports.getOracle = getOracle;
const getNamespacePDA = async (mint) => {
    return await anchor_1.web3.PublicKey.findProgramAddress([Buffer.from(namespace_1.PREFIX), mint.toBuffer()], programIds_1.NAMESPACE_ID);
};
exports.getNamespacePDA = getNamespacePDA;
const getItemPDA = async (mint, index) => {
    return await anchor_1.web3.PublicKey.findProgramAddress([Buffer.from(item_1.PREFIX), mint.toBuffer(), index.toBuffer("le", 8)], programIds_1.ITEM_ID);
};
exports.getItemPDA = getItemPDA;
const getPlayerPDA = async (mint, index) => {
    return await anchor_1.web3.PublicKey.findProgramAddress([Buffer.from(player_1.PREFIX), mint.toBuffer(), index.toBuffer("le", 8)], programIds_1.PLAYER_ID);
};
exports.getPlayerPDA = getPlayerPDA;
const getItemActivationMarker = async (args) => {
    return await anchor_1.web3.PublicKey.findProgramAddress([
        Buffer.from(item_1.PREFIX),
        args.itemMint.toBuffer(),
        args.index.toBuffer("le", 8),
        args.usageIndex.toBuffer("le", 8),
        args.amount.toBuffer("le", 8),
        Buffer.from(item_1.MARKER),
    ], programIds_1.ITEM_ID);
};
exports.getItemActivationMarker = getItemActivationMarker;
const getCraftItemCounter = async (args) => {
    return await anchor_1.web3.PublicKey.findProgramAddress([
        Buffer.from(item_1.PREFIX),
        args.itemClassMint.toBuffer(),
        args.classIndex.toBuffer("le", 8),
        args.newItemMint.toBuffer(),
        args.craftEscrowIndex.toBuffer("le", 8),
        args.craftItemIndex.toBuffer("le", 8),
        args.craftItemMint.toBuffer(),
        Buffer.from(args.componentScope),
    ], programIds_1.ITEM_ID);
};
exports.getCraftItemCounter = getCraftItemCounter;
const getCraftItemEscrow = async (args) => {
    return await anchor_1.web3.PublicKey.findProgramAddress([
        Buffer.from(item_1.PREFIX),
        args.itemClassMint.toBuffer(),
        args.classIndex.toBuffer("le", 8),
        args.payer.toBuffer(),
        args.newItemMint.toBuffer(),
        args.craftItemToken.toBuffer(),
        args.craftEscrowIndex.toBuffer("le", 8),
        args.craftIndex.toBuffer("le", 8),
        args.craftItemMint.toBuffer(),
        args.amountToMake.toBuffer("le", 8),
        args.amountToContributeFromThisContributor.toBuffer("le", 8),
        Buffer.from(args.componentScope),
    ], programIds_1.ITEM_ID);
};
exports.getCraftItemEscrow = getCraftItemEscrow;
const getItemEscrow = async (args) => {
    return await anchor_1.web3.PublicKey.findProgramAddress([
        Buffer.from(item_1.PREFIX),
        args.itemClassMint.toBuffer(),
        args.classIndex.toBuffer("le", 8),
        args.payer.toBuffer(),
        args.newItemMint.toBuffer(),
        args.newItemToken.toBuffer(),
        args.craftEscrowIndex.toBuffer("le", 8),
        args.amountToMake.toBuffer("le", 8),
        Buffer.from(args.componentScope),
    ], programIds_1.ITEM_ID);
};
exports.getItemEscrow = getItemEscrow;
const getMetadata = async (mint) => {
    return (await anchor_1.web3.PublicKey.findProgramAddress([
        Buffer.from("metadata"),
        programIds_1.TOKEN_METADATA_PROGRAM_ID.toBuffer(),
        mint.toBuffer(),
    ], programIds_1.TOKEN_METADATA_PROGRAM_ID))[0];
};
exports.getMetadata = getMetadata;
const getEdition = async (mint) => {
    return (await anchor_1.web3.PublicKey.findProgramAddress([
        Buffer.from("metadata"),
        programIds_1.TOKEN_METADATA_PROGRAM_ID.toBuffer(),
        mint.toBuffer(),
        Buffer.from("edition"),
    ], programIds_1.TOKEN_METADATA_PROGRAM_ID))[0];
};
exports.getEdition = getEdition;
